src/internal_telemetry/allocations/allocator/token.rs

```diff
@@ -0,0 +1,146 @@
+use std::{
+    cell::RefCell,
+    num::NonZeroUsize,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+use tracing::Span;
+
+use super::tracing::WithAllocationGroup;
+
+use crate::internal_telemetry::allocations::allocator::stack::GroupStack;
+
+thread_local! {
+    /// The currently executing allocation token.
+    ///
+    /// Any allocations which occur on this thread will be associated with whichever token is
+    /// present at the time of the allocation.
+    pub(crate) static LOCAL_ALLOCATION_GROUP_STACK: RefCell<GroupStack<256>> =
+        RefCell::new(GroupStack::new());
+}
+
+/// The identifier that uniquely identifiers an allocation group.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct AllocationGroupId(NonZeroUsize);
+
+impl AllocationGroupId {
+    /// The group ID used for allocations which are not made within a registered allocation group.
+    pub const ROOT: Self = Self(unsafe { NonZeroUsize::new_unchecked(1) });
+
+    pub(super) const fn from_raw_unchecked(raw_group_id: usize) -> Self {
+        unsafe { Self(NonZeroUsize::new_unchecked(raw_group_id)) }
+    }
+
+    /// Gets the integer representation of this group ID.
+    #[must_use]
+    pub const fn as_raw(self) -> usize {
+        self.0.get()
+    }
+
+    /// Registers an allocation group ID.
+    ///
+    /// This group ID uniquely identifies a given allocation group, and is the means by which to
+    /// distinguish allocator events between various allocation groups.
+    ///
+    /// Group IDs must be attached to a [`Span`][tracing::Span] in order to become active,
+    /// associating allocations and deallocations within an active span as being attached to the
+    /// given allocation group.
+    pub fn register() -> Option<AllocationGroupId> {
+        static GROUP_ID: AtomicUsize = AtomicUsize::new(AllocationGroupId::ROOT.0.get() + 1);
+
+        let group_id = GROUP_ID.fetch_add(1, Ordering::Relaxed);
+
+        if group_id != usize::MAX {
+            let group_id = NonZeroUsize::new(group_id).expect("bug: GROUP_ID overflowed");
+            Some(AllocationGroupId(group_id))
+        } else {
+            None
+        }
```

Comment:
Simpler with `bool::then`:
```suggestion
        (group_id != usize::MAX).then(|| {
            let group_id = NonZeroUsize::new(group_id).expect("bug: GROUP_ID overflowed");
            AllocationGroupId(group_id)
        })
```